src/Database/LSMTree/Internal/WriteBufferBlobs.hs

```diff
@@ -169,22 +189,47 @@ mkWeakBlobRef (DeRef WriteBufferBlobs {blobFile}) blobspan =
       weakBlobRefSpan = blobspan
     }
 
-
 -- | A mutable file offset, suitable to share between threads.
+#if MIN_VERSION_base(4,16,1)
+newtype FilePointer m = FilePointer (PrimVar (PrimState m) Word)
+#else
 newtype FilePointer m = FilePointer (PrimVar (PrimState m) Int)
---TODO: this would be better as Word64
--- this will limit to 31bit file sizes on 32bit arches
+#endif
+
 
 instance NFData (FilePointer m) where
   rnf (FilePointer var) = var `seq` ()
 
 {-# SPECIALISE newFilePointer :: IO (FilePointer IO) #-}
 newFilePointer :: PrimMonad m => m (FilePointer m)
-newFilePointer = FilePointer <$> P.newPrimVar 0
+newFilePointer = FilePointer <$> P.newPinnedPrimVar 0
+-- It is very important that a /pinned/ PrimVar is used here,
+-- as it is a pre-requisite for the operations performed in @updateFilePointer@!
 
-{-# SPECIALISE updateFilePointer :: FilePointer IO -> Int -> IO Word64 #-}
+{-# SPECIALISE updateFilePointer :: FilePointer IO -> Int    -> IO Word64 #-}
+{-# SPECIALISE updateFilePointer :: FilePointer IO -> Word   -> IO Word64 #-}
+{-# SPECIALISE updateFilePointer :: FilePointer IO -> Word64 -> IO Word64 #-}
 -- | Update the file offset by a given amount and return the new offset. This
 -- is safe to use concurrently.
---
-updateFilePointer :: PrimMonad m => FilePointer m -> Int -> m Word64
-updateFilePointer (FilePointer var) n = fromIntegral <$> P.fetchAddInt var n
+updateFilePointer :: (Integral i, PrimMonad m) => FilePointer m -> i -> m Word64
+#if MIN_VERSION_base(4,16,1)
+updateFilePointer (FilePointer (PrimVar (MutablePrimArray arr))) n =
+  let !(W# delta) = fromIntegral n
+  in  fmap fromIntegral . primitive $ \s ->
+        case fetchAddWordAddr# (mutableByteArrayContents# arr) delta s of
+          (# s', result #) -> (# s', W# result #)
+#else
+updateFilePointer (FilePointer var) n =
+  -- Bad support here for GHC-8.10.7
+  -- Ideally, drop support for GHC-8.10.7 when reasonably possible.
+  -- TODO: Remove hack after GHC-8.10.7 is antiquated!
+  let !delta = fromIntegral n
+
+      -- Rely on the machine's binary reprentation of Int forming an additive ring
+      -- which is homomorphic with Word64 (such as 2's compliment); i.e.:
+      -- let limit = toEnum (maxBound :: Int) :: Word
+      -- in  limit + x === coerce ( (maxBound :: Int) + x )
+      convert :: Int -> Word64
+      convert = unsafeCoerce
+  in  convert <$> fetchAddInt var delta
+#endif
```

Comment:
We can rely on `Int`s being 64 bits, because we require that from our users (see `README.md`). With that assumption, we can probably just use `fromIntegral <$> P.fetchAddInt var n`

Reply:
Hmm... in that case I think you are right. 

This is a case of *"6` hours of coding can save you 5 minutes of reading the documentation"*, haha.

We can mark this PR and "will not merge" and stick with the existing implementation, maybe adding a comment reminding the reader that we can assume 64-bit `Int`s.

Reply:
Sure, let's do that. Maybe you can repurpose this PR for removing the TODO?